File: app/src/main/java/me/blog/korn123/easyphotomap/activities/FileExplorerActivity.java
package me.blog.korn123.easyphotomap.activities;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Typeface;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.HorizontalScrollView;
import android.widget.ListView;
import android.widget.TextView;

import com.beardedhen.androidbootstrap.TypefaceProvider;

import org.apache.commons.io.FilenameUtils;
import org.apache.commons.lang.StringUtils;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import me.blog.korn123.easyphotomap.R;
import me.blog.korn123.easyphotomap.adapters.ExplorerItemAdapter;
import me.blog.korn123.easyphotomap.constants.Constant;
import me.blog.korn123.easyphotomap.helper.RegistrationThread;
import me.blog.korn123.easyphotomap.models.FileItem;
import me.blog.korn123.easyphotomap.utils.CommonUtils;

/**
 * Created by CHO HANJOONG on 2016-07-16.
 */
public class FileExplorerActivity extends AppCompatActivity {

    private String current;
    private ArrayList<FileItem> listFileItem;
    private ArrayList<FileItem> listDirectoryEntity;
    private ViewGroup viewGroup;
    private ProgressDialog progressDialog;
    private ArrayAdapter<FileItem> adapter;

    @BindView(R.id.filelist)
    public ListView fileList;

    @BindView(R.id.scrollView)
    public HorizontalScrollView scrollView;

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        TypefaceProvider.registerDefaultIconSets();
        setContentView(R.layout.activity_file_explorer);
        ButterKnife.bind(this);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle(getString(R.string.file_explorer_activity_title));
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        listFileItem = new ArrayList<>();
        listDirectoryEntity = new ArrayList<>();
        current = Constant.CAMERA_DIRECTORY;
        ((TextView)findViewById(R.id.registerDirectory)).setTypeface(Typeface.DEFAULT);

        adapter = new ExplorerItemAdapter(this, this, R.layout.item_file_explorer, this.listFileItem);
        fileList.setAdapter(adapter);
        viewGroup = (ViewGroup)findViewById(R.id.pathView);
        scrollView = (HorizontalScrollView)findViewById(R.id.scrollView);

        AdapterView.OnItemClickListener mItemClickListener =
                new AdapterView.OnItemClickListener() {
                    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                        FileItem thumbnailEntity = (FileItem)parent.getAdapter().getItem(position);
                        String fileName = thumbnailEntity.getFileName();

                        if (fileName.startsWith("[") && fileName.endsWith("]")) {
                            fileName = fileName.substring(1, fileName.length()-1);
                        }

                        String path = current + "/" + fileName;
                        File f = new File(path);

                        if (f.isDirectory()) {
                            current = path;
                            refreshFiles();
                        } else {
                            if (!new File(Constant.WORKING_DIRECTORY).exists()) {
                                new File(Constant.WORKING_DIRECTORY).mkdirs();
                            }
                            PositiveListener positiveListener = new PositiveListener(FileExplorerActivity.this, FileExplorerActivity.this, FilenameUtils.getName(path) + ".origin", path);
                            CommonUtils.showAlertDialog(FileExplorerActivity.this, getString(R.string.file_explorer_message7), FileExplorerActivity.this, path, positiveListener);
                        }
                    }
                };
        fileList.setOnItemClickListener(mItemClickListener);
        refreshFiles();
    }

    @Override
    protected void onResume() {
        super.onResume();
        adapter.notifyDataSetChanged();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @OnClick({R.id.registerDirectory})
    public void buttonClick(View view) {
        switch (view.getId()) {
            case R.id.registerDirectory:
                if (listFileItem.size() - listDirectoryEntity.size() < 1) {
                    CommonUtils.showAlertDialog(this, getString(R.string.file_explorer_message9));
                } else {
                    PositiveListener positiveListener = new PositiveListener(FileExplorerActivity.this, FileExplorerActivity.this, null, null);
                    CommonUtils.showAlertDialog(FileExplorerActivity.this, getString(R.string.file_explorer_message11) , FileExplorerActivity.this, positiveListener);
                }
                break;
        }
    }

    void refreshFiles() {
        String[] arrayPath = StringUtils.split(current, "/");
        viewGroup.removeViews(0, viewGroup.getChildCount());
        String currentPath = "";
        int index = 0;
        for (String path : arrayPath) {
            currentPath += ("/" + path);
            final String targetPath = currentPath;
            TextView textView = new TextView(this);
            if(index < arrayPath.length - 1) {
                textView.setText(path + "  >  ");
            } else {
                textView.setText(path);
            }

            if (StringUtils.equals(arrayPath[arrayPath.length - 1], path)) {
                textView.setTypeface(Typeface.DEFAULT, Typeface.BOLD);
                textView.setTextColor(ContextCompat.getColor(FileExplorerActivity.this, R.color.colorPrimary));
            } else {
                textView.setTypeface(Typeface.DEFAULT);
                textView.setTextColor(ContextCompat.getColor(FileExplorerActivity.this, R.color.defaultFont));
            }
            textView.setTextSize(TypedValue.COMPLEX_UNIT_DIP, 16);
            textView.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.MATCH_PARENT));
            textView.setGravity(Gravity.CENTER_VERTICAL);
            textView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    current = targetPath;
                    refreshFiles();
                }
            });
            viewGroup.addView(textView);
            index++;
        }
        scrollView.postDelayed(new Runnable() {
            public void run() {
                scrollView.fullScroll(HorizontalScrollView.FOCUS_RIGHT);
            }
        }, 100L);
        new RefreshThread().start();
    }

    @Override
    public void onBackPressed() {
        CommonUtils.showAlertDialog(FileExplorerActivity.this, getString(R.string.file_explorer_message12), new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                finish();
            }
        }, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
            }
        });
    }

    public class PositiveListener {
        Context context;
        Activity activity;
        String fileName;
        String path;

        PositiveListener(Context context, Activity activity, String fileName, String path) {
            this.context = context;
            this.activity = activity;
            this.fileName = fileName;
            this.path = path;
        }

        public void register() {
            if (fileName != null && path != null) {
                progressDialog = ProgressDialog.show(FileExplorerActivity.this, getString(R.string.file_explorer_message5), getString(R.string.file_explorer_message6));
                Thread registerThread = new RegistrationThread(context, activity, progressDialog, fileName, path);
                registerThread.start();
            } else {
                Intent batchIntent = new Intent(FileExplorerActivity.this, BatchPopupActivity.class);
                ArrayList<String> listImagePath = new ArrayList<>();
                for (int i = listDirectoryEntity.size(); i < listFileItem.size(); i++) {
                    listImagePath.add(listFileItem.get(i).getImagePath());
                }
                batchIntent.putStringArrayListExtra("listImagePath", listImagePath);
                startActivity(batchIntent);
            }
        }
    }

    class RefreshThread extends Thread {
        @Override
        public void run() {
            listFileItem.clear();
            listDirectoryEntity.clear();
            File current = new File(FileExplorerActivity.this.current);
            String[] files = current.list();
            if (files != null) {
                for (int i = 0; i < files.length;i++) {
                    FileItem thumbnailEntity = new FileItem();
                    String path = FileExplorerActivity.this.current + "/" + files[i];
                    String name = "";
                    File f = new File(path);
                    if (f.isDirectory()) {
                        name = "[" + files[i] + "]";
                        thumbnailEntity.setImagePath(name);
                        thumbnailEntity.isDirectory = true;
                        listDirectoryEntity.add(thumbnailEntity);
                    } else {
                        name = files[i];
                        String extension = FilenameUtils.getExtension(name).toLowerCase();
                        if (!extension.matches("jpg|jpeg")) continue;
                        thumbnailEntity.setImagePath(path);
                        listFileItem.add(thumbnailEntity);
                    }
                }
            }

            if (CommonUtils.loadBooleanPreference(FileExplorerActivity.this, "enable_reverse_order")) {
                Collections.sort(listDirectoryEntity, Collections.reverseOrder());
                Collections.sort(listFileItem, Collections.reverseOrder());
            } else {
                Collections.sort(listDirectoryEntity);
                Collections.sort(listFileItem);
            }
            listFileItem.addAll(0, listDirectoryEntity);

            new Handler(Looper.getMainLooper()).post(new Runnable() {
                @Override
                public void run() {
                    adapter.notifyDataSetChanged();
                    fileList.setSelection(0);
                }
            });
        }
    }

}

File: app/src/main/java/me/blog/korn123/easyphotomap/activities/FileExplorerActivity.kt
package me.blog.korn123.easyphotomap.activities

import android.app.Activity
import android.app.AlertDialog
import android.app.ProgressDialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.graphics.Typeface
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.support.v4.content.ContextCompat
import android.util.TypedValue
import android.view.*
import android.widget.ArrayAdapter
import android.widget.HorizontalScrollView
import android.widget.TextView
import kotlinx.android.synthetic.main.activity_file_explorer.*
import me.blog.korn123.easyphotomap.R
import me.blog.korn123.easyphotomap.adapters.ExplorerItemAdapter
import me.blog.korn123.easyphotomap.constants.Constant
import me.blog.korn123.easyphotomap.dialogs.ChangeSortingDialog
import me.blog.korn123.easyphotomap.extensions.config
import me.blog.korn123.easyphotomap.helper.RegistrationThread
import me.blog.korn123.easyphotomap.models.FileItem
import me.blog.korn123.easyphotomap.utils.CommonUtils
import me.blog.korn123.easyphotomap.utils.DialogUtils
import org.apache.commons.io.FilenameUtils
import org.apache.commons.lang.StringUtils
import java.io.File
import java.util.*

/**
 * Created by CHO HANJOONG on 2016-07-16.
 */
class FileExplorerActivity : SimpleActivity() {

    private val mListFile: ArrayList<FileItem> = arrayListOf()
    private val mListDirectory: ArrayList<FileItem> = arrayListOf()
    private var mCurrent: String? = null
    private var mProgressDialog: ProgressDialog? = null
    private var mAdapter: ArrayAdapter<FileItem>? = null

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_file_explorer)

        setSupportActionBar(toolbar)
        supportActionBar?.run {
            title = getString(R.string.file_explorer_activity_title)
            setDisplayHomeAsUpEnabled(true)
        }

        mCurrent = Constant.CAMERA_DIRECTORY
        mAdapter = ExplorerItemAdapter(this, this, R.layout.item_file_explorer, this.mListFile)
        fileList.adapter = mAdapter

        fileList.setOnItemClickListener { parent, _, position, _ ->
            val thumbnailEntity = parent.adapter.getItem(position) as FileItem
            var fileName = thumbnailEntity.fileName

            if (fileName.startsWith("[") && fileName.endsWith("]")) {
                fileName = fileName.substring(1, fileName.length - 1)
            }

            val path = mCurrent + "/" + fileName
            val f = File(path)

            if (f.isDirectory) {
                mCurrent = path
                refreshFiles()
            } else {
                if (!File(Constant.WORKING_DIRECTORY).exists()) {
                    File(Constant.WORKING_DIRECTORY).mkdirs()
                }
                val positiveListener = PositiveListener(this@FileExplorerActivity, this@FileExplorerActivity, FilenameUtils.getName(path) + ".origin", path)
                DialogUtils.showAlertDialog(
                        this@FileExplorerActivity,
                        getString(R.string.file_explorer_message7),
                        this@FileExplorerActivity,
                        path,
                        positiveListener
                )
            }
        }

        refreshFiles()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.file_explorer, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
            R.id.sort -> showSortingDialog()
            R.id.batchModeA -> {
                val positiveListener = DialogInterface.OnClickListener { _, _ ->
                    val batchIntent = Intent(this@FileExplorerActivity, BatchPopupActivity::class.java)
                    val listImagePath = arrayListOf<String>()
                    File(mCurrent).listFiles().map { file ->
                        if (file.absoluteFile.extension.toLowerCase().matches("jpg|jpeg".toRegex())) {
                            listImagePath.add(file.absolutePath)
                        }
                    }
                    batchIntent.putStringArrayListExtra("listImagePath", listImagePath)
                    startActivity(batchIntent)
                    return@OnClickListener
                }

                AlertDialog.Builder(this@FileExplorerActivity).apply {
                    setMessage(getString(R.string.file_explorer_message11))
                    setPositiveButton(getString(R.string.confirm), positiveListener)
                    setNegativeButton(getString(R.string.cancel), null)
                }.create().show()
            }
            R.id.batchModeB -> {
                val positiveListener = DialogInterface.OnClickListener { _, _ ->
                    val batchIntent = Intent(this@FileExplorerActivity, BatchPopupActivity::class.java)
                    batchIntent.putExtra("currentPath", mCurrent)
                    startActivity(batchIntent)
                    return@OnClickListener
                }

                AlertDialog.Builder(this@FileExplorerActivity).apply {
                    setMessage(getString(R.string.file_explorer_message13))
                    setPositiveButton(getString(R.string.confirm), positiveListener)
                    setNegativeButton(getString(R.string.cancel), null)
                }.create().show()
            }
        }
        return true
    }

    override fun onBackPressed() {
        DialogUtils.showAlertDialog(
                this@FileExplorerActivity,
                getString(R.string.file_explorer_message12),
                DialogInterface.OnClickListener { _, _ -> finish() },
                DialogInterface.OnClickListener { _, _ -> }
        )
    }

    private fun showSortingDialog() {
        ChangeSortingDialog(this, true, false) {
//            if (config.directorySorting and SORT_BY_DATE_MODIFIED > 0 || config.directorySorting and SORT_BY_DATE_TAKEN > 0) {
//                getDirectories()
//            } else {
//                gotDirectories(mDirs, true)
//            }
            refreshFiles()
        }
    }

    private fun refreshFiles() {
        FileItem.sorting = config.directorySorting
        val arrayPath = StringUtils.split(mCurrent, "/")
        pathView.removeViews(0, pathView.childCount)
        var currentPath = ""
        var index = 0
        arrayPath.map { path ->
            currentPath += "/" + path
            val targetPath = currentPath
            val textView = TextView(this)
            if (index < arrayPath.size - 1) {
                textView.text = "$path  >  "
            } else {
                textView.text = path
            }

            if (StringUtils.equals(arrayPath[arrayPath.size - 1], path)) {
                textView.setTypeface(Typeface.DEFAULT, Typeface.BOLD)
                textView.setTextColor(ContextCompat.getColor(this@FileExplorerActivity, R.color.colorPrimary))
            } else {
                textView.typeface = Typeface.DEFAULT
                textView.setTextColor(ContextCompat.getColor(this@FileExplorerActivity, R.color.default_text_color))
            }
            textView.setTextSize(TypedValue.COMPLEX_UNIT_DIP, 16f)
            textView.layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.MATCH_PARENT)
            textView.gravity = Gravity.CENTER_VERTICAL
            textView.setOnClickListener {
                mCurrent = targetPath
                refreshFiles()
            }
            pathView.addView(textView)
            index++
        }
        progressDialog.visibility = View.VISIBLE
        RefreshThread().start()
    }

    inner class PositiveListener internal constructor(val context: Context, val activity: Activity, val fileName: String, private val path: String) {
        fun register() {
            mProgressDialog = ProgressDialog.show(this@FileExplorerActivity, getString(R.string.file_explorer_message5), getString(R.string.file_explorer_message6))
            mProgressDialog?.let { it ->
                val registerThread = RegistrationThread(context, activity, it, fileName, path)
                registerThread.start()
            }
        }
    }

    inner class RefreshThread : Thread() {
        override fun run() {
            mListFile.clear()
            mListDirectory.clear()
            val current = File(mCurrent)
            val files = current.list()
            if (files != null) {
                for (i in files.indices) {
                    val fileItem = FileItem()
                    val path = "$mCurrent/${files[i]}"
                    val name: String
                    val f = File(path)
                    if (f.isDirectory) {
                        name = "[${files[i]}]"
                        fileItem.setImagePathAndFileName(name)
                        fileItem.isDirectory = true
                        mListDirectory.add(fileItem)
                    } else {
                        name = files[i]
                        val extension = FilenameUtils.getExtension(name).toLowerCase()
                        if (!extension.matches("jpg|jpeg".toRegex())) continue
                        fileItem.setImagePathAndFileName(path)
                        try {
                            fileItem.length = f.length()
                            fileItem.takenDate = CommonUtils.getDateFromJpegMetaData(f)
                        } catch (e: Exception){
                            e.printStackTrace()
                        }

                        mListFile.add(fileItem)
                    }
                }
            }

//                if (config.enableReverseOrder) {
//                    Collections.sort(mListDirectory, Collections.reverseOrder<Any>())
//                    Collections.sort(mListFile, Collections.reverseOrder<Any>())
//                } else {
            Collections.sort(mListDirectory)
            Collections.sort(mListFile)
//                }
            mListFile.addAll(0, mListDirectory)
            
            Handler(Looper.getMainLooper()).post {
                mAdapter?.notifyDataSetChanged()
                fileList.setSelection(0)
                scrollView.fullScroll(HorizontalScrollView.FOCUS_RIGHT)
                progressDialog.visibility = View.GONE
            }
        }
    }
}
